components/widgets/users/avatars/avatars.tsx

```diff
@@ -11,34 +20,88 @@ import SimpleTooltip, {useSynchronizedImmediate} from 'components/widgets/simple
 import Avatar from 'components/widgets/users/avatar';
 
 import './avatars.scss';
+import {imageURLForUser} from 'utils/utils';
+import {RequireOnlyOne} from 'utils/conditional_types';
 
 type StylingKeys = 'size';
 
 type UserProps = ComponentProps<typeof Avatar> & {name: string};
+type AvatarProps = Omit<UserProps, StylingKeys>;
+type Participants = Array<{id: $ID<UserProfile>}>;
 
-export type Props = Pick<UserProps, StylingKeys> & {
-    users: Array<Omit<UserProps, StylingKeys>>;
-    totalUsers?: number;
-    breakAt?: number;
-}
+export type Props =
+    Pick<UserProps, StylingKeys> &
+    RequireOnlyOne<{
+        participants: Participants;
+        users: AvatarProps[];
+    }> & {
+        totalUsers?: number;
+        breakAt?: number;
+    }
 
 const OTHERS_DISPLAY_LIMIT = 99;
 
+function selectUsers(
+    state: GlobalState,
+    users: Props['users'],
+    participants: Props['participants'],
+): null | {users: AvatarProps[]; missingProfiles?: string[]} {
+    if (users) {
+        return {users};
+    }
+    if (!participants) {
+        return null;
+    }
+
+    const getDisplayName = makeGetDisplayName();
```

Comment:
This creates a memoized selector on each invocation of this method.

Reply:
I reworked this into a new selector that will only memoize the display preference, and return a memoized selector function.